Wire e2e focus/skip/mode/parallel into existing env structure

johnSchnake opened the pull request:
Removing types that are e2e specific while keeping our flags
for convenience/shorthand.

Signed-off-by: John Schnake <jschnake@vmware.com>

johnSchnake commented on cmd/sonobuoy/app/args.go:
```diff
@@ -168,62 +172,48 @@ const (
-		e2eParallelFlag, defaultMode.E2EConfig.Parallel,
-		"Specify the E2E_PARALLEL flag to the conformance tests. Overrides --mode.",
-	)
+
```
The e2econfig type will eventually go away is the intent; but the registry config flag is very different so I'll do that in another PR

johnSchnake commented on cmd/sonobuoy/app/args.go:
```diff
@@ -401,3 +391,76 @@ func AddExtractFlag(flag *bool, flags *pflag.FlagSet) {
 	)
 }
+
+type e2eFocusFlagType PluginEnvVars
```
So most of the 'added' lines are like these: we remove logical complexity throughout the codebase and for this flag-specific stuff, wire up these flag types. The flag methods of get/set/etc handle all the complexity in a very narrowly scoped point in code/time.

So in these cases, each of the focus/skip/parallel/mode, you get a new flag type which defines a custom action when you set it. These end up setting env vars but you could imagine them doing other things if needed.

johnSchnake commented on cmd/sonobuoy/app/args.go:
```diff
@@ -168,62 +172,48 @@ const (
-	)
-	e2eFlags.String(
-		e2eParallelFlag, defaultMode.E2EConfig.Parallel,
-		"Specify the E2E_PARALLEL flag to the conformance tests. Overrides --mode.",
```
Again a bit of a simplification here that I dont think loses anything. We had to nest flags within our types so we could do checking later if the flag was provided or a default value. Now we just say: let the flags figure it out.

If they provide the flags in a conflicting way (--mode=quick --e2e-focus=foo) then the flags will sort of fight over the value; last to set it wins.

I think this is reasonable and I dont see a need to bend over backwards to support wonkily-written commands like that.

vladimirvivien replied:
Agreed.
I would only suggest that if both `mode` and `e2e-focus` flags are provided together to, maybe, output a warning.

johnSchnake replied:
OK; added a warning in a persistentPrerun method now. Otherwise we have to place it in a bunch of different places. Its a simple check and would apply anywhere those flags are present/provided.

johnSchnake commented on cmd/sonobuoy/app/args_test.go:
```diff
@@ -1,89 +0,0 @@
-	ops "github.com/vmware-tanzu/sonobuoy/pkg/client"
-)
-
-func TestGetE2EConfig(t *testing.T) {
```
Mode is just a flag setting env vars; we have gen tests and smoke tests etc. No loss of coverage imo.

johnSchnake commented on cmd/sonobuoy/app/gen.go:
```diff
@@ -97,11 +97,19 @@ func GenFlagSet(cfg *genFlags, rbac RBACMode) *pflag.FlagSet {
 		return nil, errors.Wrap(err, "could not retrieve E2E config")
 	}
 
+	// TODO(jschnake) we should probably just set the dynamic plugins here, but
```
TODO was old

johnSchnake commented on test/integration/testdata/gen-config-no-flags.golden:
```diff
@@ -52,7 +52,7 @@ rules:
 data:
   config.json: |
-    {"Description":"DEFAULT","UUID":"","Version":"vSubfieldTestVersion","ResultsDir":"/tmp/sonobuoy","Resources":["apiservices","certificatesigningrequests","clusterrolebindings","clusterroles","componentstatuses","configmaps","controllerrevisions","cronjobs","customresourcedefinitions","daemonsets","deployments","endpoints","ingresses","jobs","leases","limitranges","mutatingwebhookconfigurations ...
+    {"Description":"DEFAULT","UUID":"","Version":"vSubfieldTestVersion","ResultsDir":"/tmp/sonobuoy","Resources":["apiservices","certificatesigningrequests","clusterrolebindings","clusterroles","componentstatuses","configmaps","controllerrevisions","cronjobs","customresourcedefinitions","daemonsets","deployments","endpoints","ingresses","jobs","leases","limitranges","mutatingwebhookconfigurations ...
```
Going to double check the changes here; seems the plugins returned and I'll have to make sure I understand why. Not sure if its because of some new defaults or something else.

johnSchnake replied:
In this case, it is using `sonobuoy gen --config custom` and in that custom config the plugins are null explicitly. Need to make sure we arent modifying that.

johnSchnake replied:
Hm, so it makes sense with the simplification of plugin logic. If you dont specify any plugins, we revert to our default. So maybe this is a good thing and I should just update the test to check a run which specifies a random plugin

johnSchnake commented on test/integration/testdata/subfieldTest.json:
```diff
@@ -1 +1 @@
-{"Description":"DEFAULT","UUID":"","Version":"vSubfieldTestVersion","ResultsDir":"/tmp/sonobuoy","Resources":["apiservices","certificatesigningrequests","clusterrolebindings","clusterroles","componentstatuses","configmaps","controllerrevisions","cronjobs","customresourcedefinitions","daemonsets","deployments","endpoints","ingresses","jobs","leases","limitranges","mutatingwebhookconfigurations","n ...
+{"Description":"DEFAULT","UUID":"","Version":"vSubfieldTestVersion","ResultsDir":"/tmp/sonobuoy","Resources":["apiservices","certificatesigningrequests","clusterrolebindings","clusterroles","componentstatuses","configmaps","controllerrevisions","cronjobs","customresourcedefinitions","daemonsets","deployments","endpoints","ingresses","jobs","leases","limitranges","mutatingwebhookconfigurations","n ...
```
Changing this to configpluginval for plugins in order to differentiate it.

johnSchnake commented on test/integration/testdata/gen-config-then-flags.golden:
```diff
@@ -52,7 +52,7 @@ rules:
 data:
   config.json: |
-    {"Description":"DEFAULT","UUID":"","Version":"vSubfieldTestVersion","ResultsDir":"/tmp/sonobuoy","Resources":["apiservices","certificatesigningrequests","clusterrolebindings","clusterroles","componentstatuses","configmaps","controllerrevisions","cronjobs","customresourcedefinitions","daemonsets","deployments","endpoints","ingresses","jobs","leases","limitranges","mutatingwebhookconfigurations ...
+    {"Description":"DEFAULT","UUID":"","Version":"vSubfieldTestVersion","ResultsDir":"/tmp/sonobuoy","Resources":["apiservices","certificatesigningrequests","clusterrolebindings","clusterroles","componentstatuses","configmaps","controllerrevisions","cronjobs","customresourcedefinitions","daemonsets","deployments","endpoints","ingresses","jobs","leases","limitranges","mutatingwebhookconfigurations ...
```
This is what the config says; load it as instructed. TBH its a bad configuraion since in this run they're saying in the config to run `configplugival` but dont provide that plugin anywhere.

I dont think this really impacts many folks and would only do so once until they learn how pieces work together.

TLDR; I think this is ok.

johnSchnake commented on test/integration/testdata/gen-config-no-flags.golden:
```diff
@@ -52,7 +52,7 @@ rules:
 data:
   config.json: |
-    {"Description":"DEFAULT","UUID":"","Version":"vSubfieldTestVersion","ResultsDir":"/tmp/sonobuoy","Resources":["apiservices","certificatesigningrequests","clusterrolebindings","clusterroles","componentstatuses","configmaps","controllerrevisions","cronjobs","customresourcedefinitions","daemonsets","deployments","endpoints","ingresses","jobs","leases","limitranges","mutatingwebhookconfigurations ...
+    {"Description":"DEFAULT","UUID":"","Version":"vSubfieldTestVersion","ResultsDir":"/tmp/sonobuoy","Resources":["apiservices","certificatesigningrequests","clusterrolebindings","clusterroles","componentstatuses","configmaps","controllerrevisions","cronjobs","customresourcedefinitions","daemonsets","deployments","endpoints","ingresses","jobs","leases","limitranges","mutatingwebhookconfigurations ...
```
same msg:

This is what the config says; load it as instructed. TBH its a bad configuraion since in this run they're saying in the config to run `configplugival` but dont provide that plugin anywhere.

I dont think this really impacts many folks and would only do so once until they learn how pieces work together.

TLDR; I think this is ok.

vladimirvivien commented on cmd/sonobuoy/app/args.go:
```diff
@@ -401,3 +395,84 @@ func AddExtractFlag(flag *bool, flags *pflag.FlagSet) {
+type e2eFocusFlagType PluginEnvVars
+
+func (i *e2eFocusFlagType) String() string {
+	if i != nil && (map[string]map[string]string)(*i)["e2e"] != nil {
```
Suggestion: make this `(map[string]map[string]string)` a named type so your type conversion is simpler

```
type FocusFlag map[string]map[string]string

func (i *e2eFocusFlagType) String() string {
	if i != nil && FocusFlag(*i)["e2e"] != nil {
```

johnSchnake replied:
Its actually already a named type so that was easy to swap. For some reason I thought when indexing it I had to use the underlying map type, but you're right

vladimirvivien reviewed (commented):
Left couple of comments/suggestions.
Ping me if ready for approval.

johnSchnake commented:
I also reworked the boilerplate a bit so I trimmed down that code so the focus/skip/parallel all reuse the same code path. Then I also included the ssh-user in this set of changes because it was exactly the same, just changed an env var.

vladimirvivien reviewed (approved):
/lgtm
